extension/src/pipeline/index.ts

```diff
@@ -0,0 +1,29 @@
+import { join } from 'path'
+import { appendFileSync, writeFileSync } from 'fs-extra'
+import { PipelineData } from './data'
+import { DeferredDisposable } from '../class/deferred'
+import { InternalCommands } from '../commands/internal'
+import { TEMP_DAG_FILE } from '../cli/dvc/constants'
+
+export class Pipeline extends DeferredDisposable {
+  private readonly dvcRoot: string
+  private readonly internalCommands: InternalCommands
+  private readonly data: PipelineData
+
+  constructor(dvcRoot: string, internalCommands: InternalCommands) {
+    super()
+    this.dvcRoot = dvcRoot
+    this.internalCommands = internalCommands
+    this.data = this.dispose.track(new PipelineData(dvcRoot, internalCommands))
+
+    this.dispose.track(
+      this.data.onDidUpdate(data =>
+        writeFileSync(join(this.dvcRoot, TEMP_DAG_FILE), data)
+      )
+    )
+  }
+
+  public forceRerender() {
+    return appendFileSync(join(this.dvcRoot, TEMP_DAG_FILE), '\n')
```

Comment:
[F] I tried a lot of things to make this work and this is actually the only working solution.

Reply:
for example, trying to listen to `window.onDidChangeVisibleTextEditors` doesn't work because the markdown preview editor is not a standard text editor so all of the associated events return `undefined`.